### PR TITLE
Added new data set endpoints

### LINKED ISSUE #33: Uploaded Data Unit Test [References: #15]
Unit Test [References: #15]

### LINKED ISSUE #140: [Backend] - Endpoint to load table in Saved Data
As a developer, I would like send a query consisting of the account users unique identifier to retrieve their saved datasets/graphs.

**Tasks Involved:**

- [x] Creating an endpoint accessible from frontend
- [x] Tests of the method against the database

**Acceptance Criteria:**
Should send a GET requests to a localhost:4000/datasets and retrieve found datasets back in .json format .

**Risk** - High
**Priority** - High

Notes (to be reviewed by @TiraFire):
	"On Graph page, once data is loaded, there is a heart button to save it in some table"
	"Many to many relations table most likely to be used"

### LINKED ISSUE #143: [Backend] - Endpoint to load table in Uploaded Datasets
As a developer, I would like send a query consisting of the account users unique identifier to retrieve their uploaded datasets.

**Tasks Involved:**

- [x] Creating an endpoint accessible from frontend
- [x] Tests of the method against the database

**Acceptance Criteria:**
Should send a GET requests to a localhost:4000/datasets and retrieve found datasets back in .json format .

**Risk** - High
**Priority** - High

### PR DESCRIPTION
**Pull request by silent-po2**
Added new data set endpoints to allow for searches of a user's favorited/saved data sets and for finding all the data sets that a given user uploaded. #140 #143 
Also added was support for author last name only searches, versus requiring both a first and last name. 
The names of these various backend files was also tweaked to better clarify their purpose and association.

### CODE REVIEW

File: Server/src/controllers/getDatasetController.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+     * 
+     * @param request
+     * An object containing the request information and parameters: Request 
+     * @param response 
+     * An object containing a response: Response
+     */
+    async createRequestForData(request: Request, response: Response) {
+        let validateData = this.validateInputData(request)
+        if (!validateData) {
+            response.status(500).send("Invalid search params entered");
```
Comment (GD-Dheer) [inline Server/src/controllers/getDatasetController.ts]: Error 500 is a server related error, if the params are invalid we should either throw a badRequest error or instead return with error 400 instead

File: Server/src/controllers/getDatasetController.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     * send it to the getDataService to acquire an array of data sets that were favorited by this user ID, and
+     * then return a response with the array of data sets. If the request is invalid, or some other problem arrises, 
+     * it will throw an error.
+     * 
+     * @param request
+     * An object containing the request information and parameters: Request 
+     * @param response 
+     * An object containing a response: Response
+     */
+    async createRequestForUserFavorites(request: Request, response: Response) {
```
Comment (GD-Dheer) [inline Server/src/controllers/getDatasetController.ts]: What is the difference here between userFavorites and userUploads?

File: Server/src/controllers/getDatasetController.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        let requestParam = request.params.favoritesOf;
+        let userId: number = +requestParam;
+        if (isNaN(userId)) {
+            response.status(500).send("Invalid search params entered");
```
Comment (GD-Dheer) [inline Server/src/controllers/getDatasetController.ts:82]: 400 instead

File: Server/src/migrations/SeedDatabase.ts [hunk 4]
```diff
@@ -17,6 +18,27 @@ export class SeedDatabase1608609071666 implements MigrationInterface {
...
+
+    let account = new Accounts();
+    account.id = 1;
+    account.email = "test@email.com";
+    account.password = "password";
+    account.firstName = "Crystal";
+    account.lastName = "Grail";
+    account.dateOfBirth = new Date(2000 - 12 - 17);
+    account.organizationName = "Databoom Org";
+    account.createdAt;
```
Comment (GD-Dheer) [inline Server/src/migrations/SeedDatabase.ts]: Do we need to specify these?

File: Server/src/models/entities/Accounts.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    * Dataset table together. This will generate a new third table that contains
+    * Foreign Keys of linked Dataset and Accounts IDs.
+    * The 'material => material.datasets' line is added for use in query building
+    * for it defines the direction of the "link" with the Material table
+    * 
+    * Specifically this is being used to track the 'favorited' datasets of a user
+    */
+    @ManyToMany(type => Dataset, dataset => dataset.accounts)
+    @JoinTable()
+    datasets: Dataset[];
```
Comment (GD-Dheer) [inline Server/src/models/entities/Accounts.ts:49]: A+

File: Server/src/models/entities/Dataset.ts [hunk 6]
```diff
@@ -72,6 +73,30 @@ export class Dataset {
...
+    /* 
+    * This particular ManyToOne relation is being used to track which account uploaded
+    * the dataset in question
+    */
+    @Column({ nullable: true })
+    uploaderId: number
+
+    @ManyToOne(type => Accounts)
+    @JoinColumn()
+    uploader?: Accounts
```
Comment (GD-Dheer) [inline Server/src/models/entities/Dataset.ts:98]: Why is this optional?

File: Server/src/routes/getDatasetRouter.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * This file contains the route for a call to query or obtain one or more data sets. 
+ * If an API call is made to /dataset* then the request is routed to the getDataController
+ * to continue processing of the request.
+ */
+
+let router = Router();
+
+router.get('/dataset/uploadedBy/:uploadedBy', (request: Request, response: Response) => {
+    let getDataControllerObject = new getDataController();
```
Comment (GD-Dheer) [inline Server/src/routes/getDatasetRouter.ts]: Maybe a better way could be to create the new object like router and only have to create 1 obj. Then call the respective methods under the router

File: Server/src/tests/getDatasetService.spec.ts [hunk 8]
```diff
@@ -95,6 +95,22 @@ describe('data service test', () => {
     done()
   });
 
+  test('Feeds last name of Marsh andexpects to see the author named Stanley P. Marsh', async done => {
```
Comment (GD-Dheer) [inline Server/src/tests/getDatasetService.spec.ts]: Small spacing here between andexpected

**Review by GD-Dheer — COMMENTED** [review]
Looks great! A few comments were left to address. Added tests! A+

File: Server/src/controllers/getDatasetController.ts (hunk 2 above)
Comment (silent-po2) [reply Server/src/controllers/getDatasetController.ts]: One is referring to data sets that a user wants to keep track of and the other is specifically _who_ uploaded a particular data set

File: Server/src/migrations/SeedDatabase.ts (hunk 4 above)
Comment (silent-po2) [reply Server/src/migrations/SeedDatabase.ts]: Resolved as of e44fbb8

File: Server/src/models/entities/Dataset.ts (hunk 6 above)
Comment (silent-po2) [reply Server/src/models/entities/Dataset.ts:98]: It is optional due to data sets uploaded from the old web site, which will not be affixed to a particular user.

File: Server/src/controllers/getDatasetController.ts (hunk 2 above)
Comment (TiraFire) [reply Server/src/controllers/getDatasetController.ts]: As he mentioned, the userUploads is used for the Uploaded table which shows the user all the datasets they uploaded to the database. Versus a table referencing all the datasets from the database he wants to keep an eye on or have easy access to.

File: Server/src/routes/getDatasetRouter.ts (hunk 7 above)
Comment (TiraFire) [reply Server/src/routes/getDatasetRouter.ts]: Could you explain this another way? I have a hard time understanding what you would like done. 👍

Comment (silent-po2) [reply Server/src/routes/getDatasetRouter.ts]: Good catch, will do that!

File: Server/src/controllers/getDatasetController.ts (hunk 2 above)
Comment (GD-Dheer) [reply Server/src/controllers/getDatasetController.ts]: Can we update the names in that case. Its confusing to understand the difference without this explanation

**Review by TiraFire — APPROVED** [review]
Everything seems to look good.

File: Server/src/controllers/getDatasetController.ts (hunk 2 above)
Comment (silent-po2) [reply Server/src/controllers/getDatasetController.ts]: Names have been changed to userUploadedDatasets and userSavedDatsets for better clarity

**Review by GD-Dheer — APPROVED** [review]
Comments all addressed. A+

**Pull request merged**